File: haskell/FenwickTree.hs
{-# LANGUAGE BangPatterns, FlexibleContexts, Safe #-}
module FenwickTree (
  update, reduce
) where

import Control.Monad
import Control.Monad.ST
import Data.Array.ST.Safe
import Data.Bits
import Data.Int

{------------------ Fenwick tree --------------------------------}
type FenwickTree a e = a Int e

update :: (MArray a e m) => FenwickTree a e -> (e -> e) -> Int -> m ()
{-# SPECIALIZE update :: FenwickTree (STUArray s) Int32 -> (Int32 -> Int32) -> Int -> ST s () #-}
update a f !x = do
  (0, !n) <- getBounds a
  let loop !i = when (i <= n) $ readArray a i >>= (writeArray a i . f) >> loop (i .|. succ i)
  loop x

--reduce on [0, x + 1)
reduce :: (MArray a e m) => FenwickTree a e -> (b -> e -> b) -> b -> Int -> m b
{-# SPECIALIZE reduce :: FenwickTree (STUArray s) Int32 -> (Int32 -> Int32 -> Int32) -> Int32 -> Int -> ST s Int32  #-}
reduce a f !zero !x =
  let loop !i !r = if i < 0 then return r
                   else readArray a i >>= loop (pred $! i .&. succ i) . f r
  in loop x zero
{---------------------------------------------------------------}

File: kotlin/src/main/kotlin/FenwickTree.kt
class IntFenwickTree(val n: Int) {
  private val a = IntArray(n)
  fun update(x: Int, v: Int) {
    assert(x < n)
    var i = x
    while (i < n) {
      a[i] += v
      i = i or (i + 1)
    }
  }
  fun reduce(x: Int): Int {
    assert(x < n)
    var r = 0
    var i = x
    while (i >= 0) {
      r += a[i]
      i = (i and (i + 1)) - 1
    }
    return r
  }
}
